MapleServer2/Triggers/TriggerConditions.cs

```diff
@@ -132,28 +137,51 @@ public bool PvpZoneEnded(byte arg1)
         return false;
     }
 
-    public bool QuestUserDetected(int[] arg1, int[] arg2, byte[] arg3, byte arg4)
+    public bool QuestUserDetected(int[] boxes, int[] questIds, byte[] modes, byte arg4)
     {
+        QuestState mode = (QuestState) modes[0];
+        foreach (int boxId in boxes)
+        {
+            MapTriggerBox box = MapEntityStorage.GetTriggerBox(Field.MapId, boxId);
+            List<IFieldActor<Player>> players = Field.State.Players.Values.ToList();
+
+            foreach (IFieldObject<Player> player in players)
+            {
+                if (!FieldManager.IsPlayerInBox(box, player))
+                {
+                    continue;
+                }
+
+                foreach (int questId in questIds)
+                {
+                    QuestStatus quest = player.Value.QuestList.FirstOrDefault(x => x.Id == questId);
+                    if (quest is null)
+                    {
+                        return false;
+                    }
+
+                    switch (mode)
+                    {
+                        case QuestState.Started:
+                            return quest.State is QuestState.Started;
+                        case QuestState.ConditionCompleted:
+                            return quest.Condition.All(condition => condition.Completed);
+                        case QuestState.Finished:
+                            return quest.State is QuestState.Finished;
+                    }
+                }
+            }
+        }
+
         return false;
     }
 
-    public bool RandomCondition(float proc, string desc)
-    {
-        Random random = new();
-        float result = (float) random.NextDouble();
-        result *= 100;
-        return result <= proc;
-    }
+    public bool RandomCondition(float proc, string desc) => RandomProvider.Get().Next(100) <= proc;

```

Comment:
Eventually this is going to have to be reverted out of a lambda expression when we implement the desc parameter. Is it worth having it as it is right now? dunno and prob minor. just saying

Reply:
I'm going to leave like that, when it's needed it will be easy to fix anyway